### PR TITLE
Optimize ERF with MKL math function

### PR DESCRIPTION
**Pull request by TaoLv**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: tests/python/gpu/test_operator_gpu.py [hunk 1]
```diff
@@ -2121,6 +2121,69 @@ def test_context_num_gpus():
...
+            math_log(shape=shape, dtype=dtype, check_value=check_value)
+        elif op == 'erf':
+            math_erf(shape=shape, dtype=dtype, check_value=check_value)
+        elif op == 'square':
+            math_square(shape=shape, dtype=dtype, check_value=check_value)
+
+@with_seed()
+def test_math():
+    ops = ['log', 'erf', 'square']
+    check_value= True
```
Comment (eric-haibin-lin) [inline tests/python/gpu/test_operator_gpu.py:2242]: nit: space after check_value

File: tests/python/gpu/test_operator_gpu.py [hunk 2]
```diff
@@ -2121,6 +2121,69 @@ def test_context_num_gpus():
...
+    lshape = 1000
+    rshapes = [1, 10, 100, 1000, 10000]
```
Comment (eric-haibin-lin) [inline tests/python/gpu/test_operator_gpu.py]: Is there any reason why you picked 5 shapes?

File: tests/python/gpu/test_operator_gpu.py [hunk 3]
```diff
@@ -2121,6 +2121,69 @@ def test_context_num_gpus():
...
+        x_ = x.as_in_context(mx.cpu())
+        mx.nd.waitall()
+        y_ = mx.nd.erf(data=x_)
+        y_.wait_to_read()
+        assert_almost_equal(y.asnumpy(), y_.asnumpy())
+
+def math_square(shape, dtype, check_value):
+    np_x = np.random.rand(shape[0], shape[1])
+    x = mx.nd.array(np_x, dtype=dtype)
+    mx.nd.waitall()
```
Comment (eric-haibin-lin) [inline tests/python/gpu/test_operator_gpu.py]: Why do we need to frequently add waitall and wait_to_read in tests?

File: src/operator/tensor/elemwise_unary_op.h [hunk 4]
```diff
@@ -35,9 +35,10 @@
 #include "../mxnet_op.h"
 #include "../elemwise_op_common.h"
 #include "../../ndarray/ndarray_function.h"
+
 #if MSHADOW_USE_MKL == 1
```
Comment (eric-haibin-lin) [inline src/operator/tensor/elemwise_unary_op.h:39]: I have not set USE_MKL before. Just curious: is blas=MKL also tested in mxnet CI?

File: src/operator/tensor/elemwise_unary_op.h [hunk 5]
```diff
@@ -562,6 +570,50 @@ struct ReshapeLikeParam : public dmlc::Parameter<ReshapeLikeParam> {
     })                                                              \
   .add_argument("data", "NDArray-or-Symbol", "The input array.")
 
+/*! \bried MKL Unary compute. 
+ *  With this macro means mxnet compile with MKL to accelerate math function with mkl. 
+ *  Will Register FCompute with UnaryOp::MKL_Compute() to compelet the math function. 
+ */
+#define MXNET_MKL_OPERATOR_REGISTER_UNARY(__name$)                  \
```
Comment (eric-haibin-lin) [inline src/operator/tensor/elemwise_unary_op.h]: How is this different from `MXNET_OPERATOR_REGISTER_UNARY`?

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 6]
```diff
@@ -812,6 +812,26 @@ The storage type of ``fix`` output depends upon the input storage type:
...
+
+The storage type of ``square`` output depends upon the input storage type:
+
+   - square(default) = default
+   - square(row_sparse) = row_sparse
+   - square(csr) = csr
+
+)code" ADD_FILELINE)
+.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{"_backward_square"});
+#else
```
Comment (eric-haibin-lin) [inline src/operator/tensor/elemwise_unary_op_basic.cc:843]: Hmmm this does not scale to 100 ops. We need to think about better ways.

File: src/operator/tensor/elemwise_unary_op.h (hunk 4 above)
Comment (TaoLv) [reply src/operator/tensor/elemwise_unary_op.h:39]: `MSHADOW_USE_MKL` is widely used in mshadow and mxnet to indicate MKL is used as BLAS library. Yes, `USE_BLAS=mkl` is built and tested in CI:
https://github.com/apache/incubator-mxnet/blob/master/ci/docker/runtime_functions.sh#L375
https://github.com/apache/incubator-mxnet/blob/master/ci/docker/runtime_functions.sh#L553

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 6 above)
Comment (TaoLv) [reply src/operator/tensor/elemwise_unary_op_basic.cc:843]: Yes. Need revisit the design to make it scalable for both operator registration and the kernel launching everywhere in mxnet.

File: tests/python/gpu/test_operator_gpu.py (hunk 2 above)
Comment (juliusshufan) [reply tests/python/gpu/test_operator_gpu.py]: The test case is initialy to evaluate the scalability of mkl vml functions on different size of tensors, since it is not proper to track the time cost on calculation, I will modify the shape to 1d, 2d, 3d tensors

File: tests/python/gpu/test_operator_gpu.py (hunk 3 above)
Comment (juliusshufan) [reply tests/python/gpu/test_operator_gpu.py]: The test case is initially to calcuatee the time cost of the calculations on different size tensors, wwwill remove the waitall, wait_to_read fron the test case as the test case targting the correctness of the calculation.

File: src/operator/tensor/elemwise_unary_op.h (hunk 5 above)
Comment (juliusshufan) [reply src/operator/tensor/elemwise_unary_op.h]: The implementation of these two macros  are same, duplicates code removed.
